alternator/ttl.cc

```diff
@@ -113,4 +142,437 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+    return expiration_time <= now &&
+           expiration_time > now - std::chrono::years(5);
+}
+static bool is_expired(const big_decimal& expiration_time, gc_clock::time_point now) {
+    // Convert the big_decimal to an integer. If it doesn't fit, the
+    // conversion will result in the highest number that does fit.
+    // FIXME: what is expiration_time.scale()? Can it ever be not 1?
+    // Add a test that tries 12345e1 as a timestamp!
+    uint32_t t = static_cast<uint32_t>(expiration_time.unscaled_value());
+    // We assume - and the assumption turns out to be correct - that the
+    // epoch of gc_clock::time_point and the one used by the DynamoDB protocol
+    // are the same (the UNIX epoch in UTC). The resolution (seconds) is also
+    // the same.
+    return is_expired(gc_clock::time_point(gc_clock::duration(std::chrono::seconds(t))), now);
+}
+static bool is_expired(const rjson::value& expiration_time, gc_clock::time_point now) {
+    std::optional<big_decimal> n = try_unwrap_number(expiration_time);
+    if (n) {
+        return is_expired(*n, now);
+    }
+    return false;
+}
+
+// expire_item() expires an item - i.e., deletes it as appropriate for
+// expiration - with CL=QUORUM and (FIXME!) in a way Alternator Streams
+// understands it is an expiration event - not a user-initiated deletion.
+static future<> expire_item(service::storage_proxy& proxy,
+                            const service::query_state& qs,
+                            const std::vector<bytes_opt>& row,
+                            const std::vector<unsigned>& pk_columns,
+                            const std::vector<unsigned>& ck_columns,
+                            schema_ptr schema,
+                            api::timestamp_type ts) {
+    // Prepare the row key to delete
+    std::vector<bytes> exploded_pk;
+    for (unsigned c : pk_columns) {
+        const auto& row_c = row[c];
+        if (!row_c) {
+            // This shouldn't happen - all key columns must have values.
+            // But if it ever happens, let's just *not* expire the item.
+            return make_ready_future<>();
+        }
+        exploded_pk.push_back(*row_c);
+    }
+    auto pk = partition_key::from_exploded(exploded_pk);
+    mutation m(schema, pk);
+    // If there's no clustering key, a tombstone should be created directly
+    // on a partition, not on a clustering row - otherwise it will look like
+    // an open-ended range tombstone, which will crash on KA/LA sstable format.
+    // See issue #6035
+    if (ck_columns.empty()) {
+        m.partition().apply(tombstone(ts, gc_clock::now()));
+    } else {
+        std::vector<bytes> exploded_ck;
+        for (unsigned c : ck_columns) {
+            const auto& row_c = row[c];
+            if (!row_c) {
+                // This shouldn't happen - all key columns must have values.
+                // But if it ever happens, let's just *not* expire the item.
```

Comment:
ditto - logging an error message could be helpful

Reply:
Also adding a FIXME here.